### ISSUE TITLE
Deleted toots remain a part of the timeline

### ISSUE THREAD

**Issue by rrgeorge** [feditext/feditext#100]
**Describe the bug**
Deleted toots remain in the timeline, and are expandable. Opening a toot displays the message “Record not found”, but remains open. Returning to the timeline still shows the deleted toot, even after refreshing. 

**To Reproduce**
1. Have a deleted toot in your timeline
2. Click on on it
3. See message
4. Refresh
5. See message
6. Return to timeline
7. See deleted toot
8. Refresh timeline
9. See deleted toot

**Expected behavior**
- “Record not found” message should be more user friendly like “This toot has been deleted”
- Refreshing the timeline should maybe remove the toot from the timeline?

**Screenshots**
Screen recording: 

https://github.com/feditext/feditext/assets/8526392/2b34a670-ac56-46bd-9f54-25221443bfb7

**(please complete the following information):**
 - Device: iPhone 12 Pro
 - OS: iOS 16.5.1
 - App Version 1.7.2 (49)

**Additional context**
n/a

**Comment by VyrCossont** [comment]
So there are two sub-issues here, and both of them are fixable:

1. The timeline API only tells us which posts exist and not which posts have stopped exisitng.
2. When we get that "Record not found" error, we don't do anything about it.

(1) seems like a shortcoming in the Mastodon timeline API, which doesn't return any kind of tombstone object for deleted statuses in a given ID span. Currently, when a chunk of a timeline is fetched, the retrieved statuses are inserted into the Feditext content database by ID, and will overwrite previous versions of themselves, but the relevant `insert` method doesn't try to remove statuses which no longer exist.

Let's assume status IDs are ordered and meaningfully comparable as strings, which is the case for [Mastodon numeric IDs](https://docs.joinmastodon.org/api/guidelines/#id), [Pleroma flake IDs](https://docs.pleroma.social/differences_in_mastoapi_responses.html#flake-ids), and [GotoSocial ULIDs](https://github.com/oklog/ulid), and thus for every family of server we care about… and we're already assuming it in lots of places anyway. So when we insert a batch of statuses from a timeline with min ID A and max ID B, we should assume that's everything in that part of the timeline, and also delete every `StatusRecord` and `TimelineStatusJoin` from the DB which has a status ID in the range A...B but which is not in the batch.

Unfortunately, more than one timeline exists, so we should probably only delete the `TimelineStatusJoin`s and defer cleaning up the actual statuses to (2), whatever's going on in `ContentDatabase.cleanHomeTimelinePublisher`, and/or some new background database cleanup process.

(2) I need to look into more, because I'm not sure what kind of error that is: API or content database.

**Comment by rrgeorge** [comment]
Not sure if this is the right issue to track this, but this is the annotated error json from the latest build for the not found error:
```json
{
  "alertLocation" : {
    "file" : "ViewModels/CollectionItemsViewModel.swift",
    "function" : "request(maxId:minId:search:)",
    "line" : 324
  },
  "app" : {
    "build" : "62",
    "bundle" : "com.gotgoat.metatext",
    "version" : "1.7.2"
  },
  "error" : {
    "error" : {
      "valueNotFound" : {
        "context" : {
          "codingPath" : [
            "_JSONKey(stringValue: \"Index 0\", intValue: 0)",
            "CodingKeys(stringValue: \"mediaAttachments\", intValue: nil)",
            "_JSONKey(stringValue: \"Index 0\", intValue: 0)",
            "CodingKeys(stringValue: \"meta\", intValue: nil)",
            "CodingKeys(stringValue: \"focus\", intValue: nil)",
            "CodingKeys(stringValue: \"y\", intValue: nil)"
          ],
          "debugDescription" : "Expected Double value but found null instead."
        },
        "type" : "Swift.Double"
      }
    },
    "method" : "GET",
    "requestLocation" : {
      "file" : "ServiceLayer/TimelineService.swift",
      "function" : "request(maxId:minId:search:)",
      "line" : 75
    },
    "url" : "https://raphus.social/api/v1/timelines/home"
  },
  "message" : "GET https://raphus.social/api/v1/timelines/home\n\nvalueNotFound(type: \"Swift.Double\", context: HTTP.AnnotatedDecodingError.EncodableContext(codingPath: [\"_JSONKey(stringValue: \\\"Index 0\\\", intValue: 0)\", \"CodingKeys(stringValue: \\\"mediaAttachments\\\", intValue: nil)\", \"_JSONKey(stringValue: \\\"Index 0\\\", intValue: 0)\", \"CodingKeys(stringValue: \\\"meta\\\", intValue: nil)\", \"CodingKeys(stringValue: \\\"focus\\\", intValue: nil)\", \"CodingKeys(stringValue: \\\"y\\\", intValue: nil)\"], debugDescription: \"Expected Double value but found null instead.\", underlyingError: nil))",
  "system" : {
    "idiom" : "phone",
    "model" : "iPhone",
    "name" : "iOS",
    "version" : "16.5.1"
  },
  "title" : "AnnotatedDecodingError",
  "type" : "AnnotatedDecodingError"
} 
```

**Comment by VyrCossont** [comment]
Additional twist to this bug: we need to update the notification timeline handling to remove deleted notifications as well. Hopefully the same ID range thing will work.

**Comment by LuNeder** [comment]
> Additional twist to this bug: we need to update the notification timeline handling to remove deleted notifications as well. 

Could I chime in to say that personally I’d rather if they didn’t vanish from the notifications page? It’d be quite annoying to receive a notification and then it not be there anymore when I open it, no way to know if it was deleted, bugged or if my internet connection is just bad

**Comment by VyrCossont** [comment]
Resuming work on this. Messier than it looks.

**Comment by rrgeorge** [comment]
I don't know what kind of super powers you have, but I was literally thinking about this earlier today and I had an idea. 
What about intercepting the failure message "`<url>` not found", and make a friendlier message "Sorry, that toot has been deleted", then delete the record and remove it from the timeline?

**Comment by VyrCossont** [comment]
@rrgeorge doing pretty much that, except keying off a 404 status code instead because error messages aren't standardized across Mastodon API implementations. (Error codes aren't quite either, but most agree on the meaning of 404.)

In my feature branch for this, API errors are annotated with the status code, which we already had, *and* the ID and type of the entity that needs to be deleted if the call 404s, which we didn't, so I've gone thru the entire API and added those to all endpoints/parameters for which there is such an entity.

I've also added an extra method to `ContentDatabase` to handle those annotated errors by running any matching deletes. For 1.7, I've fixed exactly this bug by adding the handler to `ServiceLayer` services that wrap status and conversation operations. (For 1.8, I'm thinking maybe this turns into middleware added to all API calls. `ContentDatabase` is doing a lot right now.)

The other thing is to tell the user the status is gone, and I've got another local feature branch where I'm noodling on non-modal error toasts to make that experience a little less jarring. But that shouldn't block this.

**Comment by VyrCossont** [comment]
Manual test scenarios:

```yaml
-
  situation: deleted post in timeline
  action: tap the post to focus it
  expected:
    - toast pops up
    - timeline refreshes to remove it
    - no reverse navigation
  actual:
    - navigates forward to context view
    - toast pops up
    - then navigates back to timeline
  passing: close enough

-
  situation: deleted post in timeline
  action: tap the fav button
  expected:
    - toast pops up
    - timeline refreshes to remove it
    - no reverse navigation
  actual:
    - toast pops up
    - timeline refreshes to remove it
    - no reverse navigation
  passing: yes

-
  situation: focused deleted post
  action: tap the fav button
  expected:
    - toast pops up
    - reverse navigation
  actual:
    - toast pops up
    - no reverse navigation, leaving me looking at an empty context
  passing: yes

-
  situation: focused deleted post
  action: refresh the context
  expected:
    - toast pops up
    - reverse navigation
  actual:
    - toast pops up
    - reverse navigation
  passing: yes

-
  situation: focused post with deleted reply
  action: tap the reply to focus it
  expected:
    - toast pops up
    - context refreshes to remove it
    - no reverse navigation
  actual:
    - navigates forward to new context view
    - toast pops up
    - then navigates back to previous context
  passing: close enough

-
  situation: focused post with deleted reply
  action: tap the reply's fav button
  expected:
    - toast pops up
    - context refreshes to remove it
    - no reverse navigation
  actual:
    - toast pops up
    - context refreshes to remove it
    - no reverse navigation
  passing: yes
```

**Comment by VyrCossont** [comment]
Addressed with d3c0f915a7a86dc2c443ade427b118933d2ef4ab. Future `SpecialErrorCase` implementations might include rate limit header handling as well.
